frontend/js/src/settings/music-services/details/MusicServices.tsx

```diff
@@ -0,0 +1,257 @@
+import * as React from "react";
+
+import { capitalize } from "lodash";
+import { useLoaderData } from "react-router-dom";
+import { toast } from "react-toastify";
+import { ToastMsg } from "../../../notifications/Notifications";
+import ServicePermissionButton from "./components/ExternalServiceButton";
+
+type MusicServicesLoaderData = {
+  current_spotify_permissions: string;
+  current_critiquebrainz_permissions: string;
+  current_soundcloud_permissions: string;
+};
+
+export default function MusicServices() {
+  const loaderData = useLoaderData() as MusicServicesLoaderData;
+
+  const [permissions, setPermissions] = React.useState({
+    spotify: loaderData.current_spotify_permissions,
+    critiquebrainz: loaderData.current_critiquebrainz_permissions,
+    soundcloud: loaderData.current_soundcloud_permissions,
+  });
+
+  const handlePermissionChange = async (
+    serviceName: string,
+    newValue: string
+  ) => {
+    try {
+      const response = await fetch(
+        `/profile/music-services/${serviceName}/disconnect/`,
+        {
+          method: "POST",
+          body: JSON.stringify({ action: newValue }),
+          headers: {
+            "Content-Type": "application/json",
+          },
+        }
+      );
+
+      if (newValue === "disable") {
+        toast.success(
+          <ToastMsg
+            title="Success"
+            message={`${capitalize(
+              serviceName
+            )} integration has been disabled.`}
+          />
+        );
+
+        setPermissions((prevState) => ({
+          ...prevState,
+          [serviceName]: newValue,
+        }));
+        return;
+      }
+
+      const data = await response.json();
+      const { url } = data;
+
+      window.location.href = url;
```

Comment:
Here's the culprit for the page refresh. Probably unwanted.

Reply:
The website does not refresh, what happens in background is that the `/disconnect/` returns the `url` which is the link to the external service provider which redirects to the callback url in the backend which then reloads the page.

Right now you would be signed in to the respective service, you would have observed a reload. If you try it on incognito, then you can test the entire flow.

Reply:
OK, I read the code in profile.py, thanks for the explanation.

Now I remember why we had to set `window.location.href`: that is the only way we can show the spotify login/authorization windows.

I haven't tried it, but I wonder if there is any chance of us improving things with `redirect: 'follow'` in the fetch options. Not sure that will do anything better than whet we are currently doing.
Anyway, that's a problem for another day.